lib/evmone/eof.cpp

```diff
@@ -47,38 +71,90 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
+                if (!section_headers[DATA_SECTION].empty())
                     return EOFValidationError::multiple_data_sections;
                 state = State::section_size;
                 break;
             case CODE_SECTION:
-                if (section_headers[CODE_SECTION] != 0)
-                    return EOFValidationError::multiple_code_sections;
+            {
+                if (section_headers[TYPE_SECTION].empty())
+                    return EOFValidationError::code_section_before_type_section;
+                if (!section_headers[DATA_SECTION].empty())
+                    return EOFValidationError::data_section_before_code_section;
+                if (!section_headers[CODE_SECTION].empty())
+                    return EOFValidationError::multiple_code_sections_headers;
+                if (it == container_end)
+                    return EOFValidationError::incomplete_section_number;
+                const auto section_number_hi = *it++;
+                if (it == container_end)
+                    return EOFValidationError::incomplete_section_number;
+                const auto section_number_lo = *it++;
+                section_num = static_cast<uint16_t>((section_number_hi << 8) | section_number_lo);
+                if (section_num == 0)
+                    return EOFValidationError::zero_section_size;
                 state = State::section_size;
                 break;
+            }
             default:
                 return EOFValidationError::unknown_section_id;
             }
             break;
         }
         case State::section_size:
         {
-            const auto size_hi = *it++;
-            if (it == container_end)
-                return EOFValidationError::incomplete_section_size;
-            const auto size_lo = *it++;
-            const auto section_size = static_cast<uint16_t>((size_hi << 8) | size_lo);
-            if (section_size == 0)
-                return EOFValidationError::zero_section_size;
-
-            section_headers[section_id] = section_size;
+            if (section_id == CODE_SECTION)
+            {
+                assert(section_num > 0);  // Guaranteed by previous validation step.
+                for (size_t i = 0; i < section_num; ++i)
+                {
+                    if (it == container_end)
+                        return EOFValidationError::incomplete_section_size;
+                    const auto size_hi = *it++;
+                    if (it == container_end)
+                        return EOFValidationError::incomplete_section_size;
+                    const auto size_lo = *it++;
+                    const auto section_size = static_cast<uint16_t>((size_hi << 8) | size_lo);
+                    if (section_size == 0)
+                        return EOFValidationError::zero_section_size;
+
+                    if (section_headers[CODE_SECTION].size() == CODE_SECTION_NUMBER_LIMIT)
```

Comment:
This should be checked earlier when we're parsing number of code sections, no need to parse every code section size, if there're too many.

Reply:
@chfast @gumb0 was this comment addressed?

Reply:
Looks like it was not.

Reply:
Just fixed.